contract-exe: Better executable contracts

j-mueller opened the pull request:
This is the first step towards a new way of writing the client-side portion of our contracts. This PR only  implements the bare minimum (see below for a list of things that are still missing) but I think it's best to get some feedback on it now.

* This adds the `Contract` type of contracts that produce transactions and wait for user inputs & blockchain events.
* The `Step` type describes the inputs that the contract accepts in its current state: User inputs, changes to interesting addresses on the blockchain, and the arrival of a specific `Slot`. For each of those there is a corresponding constructor in the `Event` type. After applying an `Event` we get the next `Step` with a new set of acceptable inputs. The `Step` / `Event` types replace the trigger mechanism and the endpoints that we currently have in `WalletAPI` and in the playground.
* `Language.Plutus.Contract.Servant` turns a `PlutusContract` into a servant `Application` with two endpoints for the contract: `initialise`

### Example

See `plutus-contract-exe/examples/guessing-game/Main.hs`. Running `GET /initialise` on the guessing game server gives

```
{
  "stepEndpoints": [
    "lock"
  ],
  "stepNextSlot": null,
  "stepAddresses": [
    {
      "getAddress": "9f18eb185018510508189018ab0a1882186118ab18d0186f1840186f18f21883181a18631854184e06189818f2189c18da18e918b8186d187518c8188eff"
    }
  ],
  "stepTransactions": []
}
```

Two actions are acceptable: Calling the `lock` endpoint, and updating the contract state with a transaction that modifies the address `9f18eb18`. If we `POST` such a transaction to the `run` route then the result shows two endpoints: `lock`, as before, and `guess`, which has become available because the contract now knows about an unspent output at the game address that it can unlock.

### TO DO (future PRs)

* Static endpoints (#1115, see also (plutus-contract-exe\docs\contract-api.adoc))
* More tests
* Move the `Contract` type to `plutus-emulator`, use it in the mockchain and rewrite the use cases etc. in this style
* Try to make the `Language.Plutus.Contract.State` module nicer - less boilerplate, properties/laws, tests

j-mueller commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,86 @@
+        Emit t c'  -> Emit t (c' >>= f)
+
+instance Semigroup a => Semigroup (Contract t i a) where
+    (<>) = liftA2 (<>)
```
This waits for both sides to finish. Another option would be to select the contrat that finishes first (ie. match on the arguments and return the first `Pure` we see). This might actually be nicer - it has the benefit of not needing a `Semigroup a`, we could have `Done` as a unit and it would also work better with recursion (for contracts that have some internal state)

michaelpj replied:
So as I understand it, this waits for both to finish and then combines them? That seems reasonable. We could provide the "racing" one as the `Alternative` instance, perhaps?

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,86 @@
+    cf <*> ca = case cf of
+        Done -> Done
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> case ca of
```
Isn't this case just `fmap f ca`? Might need to write a separate `Functor` instance.

effectfully replied:
I think it's just `fmap f ca`.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,86 @@
+            Emit t c   -> Emit t (Pure f <*> c)
+            Pure a     -> Pure (f a)
+            Waiting f' -> Waiting $ \i' -> fmap f (f' i')
+        Waiting f ->
```
I feel like this case could be simpler too but I'm not clever enough to see how atm.

effectfully replied:
Hm. So having

```haskell
f :: i -> Contract t i (a -> b)
c :: Contract t i a
```

we need to produce

```haskell
res :: Contract t i b
```

We can transform `f` into this form:

```haskell
f :: i -> Contract t i a -> Contract t i b
```

and the overall thing looks like mapping a function [iover](http://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Setter.html#v:iover) a contract.

Alternatively, `f` is an instance (and reordering) of

```haskell
f :: (i -> Contract t i a) -> i -> Contract t i b
```

So the general signature is

```haskell
((i -> Contract t i a) -> i -> Contract t i b) -> Contract t i a -> Contract t i b
```

and morally this functions transforms a continuation of type `i -> Contract t i a` into some other one in a `Contract t i a`.

I'm not able to spot a nicer pattern.

michaelpj replied:
Here's my attempt at a rewrite, which I think is nicer.
```
    Emit t c  <*> ca       = Emit t (c <*> ca)
    cf        <*> Emit t c = Emit t (cf <*> c)
    Pure f    <*> ca       = f <$> ca
    Waiting f <*> ca       = Waiting $ \i -> f i <*> offer i ca
        where
            offer i (Waiting f) = f i
            offer _ c           = c
```

j-mueller replied:
That's good, it makes sense to look at the `Emit` on the right as well.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,86 @@
+    Waiting f -> (mempty, Waiting f)
+    Done -> (mempty, Done)
+    Pure a -> (mempty, Pure a)
+    Emit t c -> first (t <>) (drain c)
```
This function might be nicer in `Writer` - the instance for a monoid is basically what you're implementing here.

michaelpj replied:
Same for the `applyInput` functions, I suspect they would become much simpler.

j-mueller replied:
What exactly do you mean by that? `ContractT (Writer t)` ? or `instance MonadWriter t (Contract t)` ? I tried various combinations of `WriterT` and `Contract` and they all resulted in a lot more code

michaelpj replied:
It would have to be `ContractT (Writer t) a`, but you can do a poor-man's version by just doing `Writer t (Contract t i a)`.

I can also have a look and see if I can cash that out in actual code ;)

j-mueller replied:
Sure, have a look ;)

michaelpj replied:
https://github.com/michaelpj/plutus/commit/6a6539714a50513f17c93de554b57192ad31448a

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,86 @@
+isDone = (\case { Done -> True; _ -> False }) . snd . drain
+
+await :: t -> (i -> Maybe a) -> Contract t i a
+await a f = Emit a $ Waiting $ maybe (await a f) Pure . f
```
I find it odd that this combines emitting with awaiting. Why not add an `emit` combinator?

Also, I'd be inclined to make the type just `await :: Contract t i i`, i.e. it always gives you the input value. This is good enough - we can always turn it into some other type later, e.g.
```
x <- await maybeIToX
```
vs
```
i <- await
let x = maybeIToX i
```

Having written that,  I now see that it calls itself if the maybe-function doesn't return true (I found this quite hard to read, I think it would be better with a pattern match, tbh). I think I'd still be inclined to have something simpler, like `await :: (i -> Bool) -> Contract t i a` or something.

michaelpj replied:
The more I think about it the more I think I want
```
await :: Contract t i i
awaitUntil :: (i -> Bool) -> Contract t i i
awaitUntil p = do
    i <- await
    if p i then pure i else awaitUntil p
```

j-mueller replied:
I'll add a combinator for each of the constructors of `Contract`. 

But I still think `await` is useful as it is: The type `i` is something like `data Input = LedgerUpdate | SlotChange | Endpoint` so with `i -> Bool` we still have to pattern match on the `i` if the predicate is true and I think it's better to combine that into one. 

That said, I'm not quite happy with the emit/await structure yet. I'll see if I can make it nicer.

michaelpj replied:
Note that something like `until` is a self-contained monad combinator, and you can see various people have defined it: https://hoogle.haskell.org/?hoogle=%28a+-%3E+Bool%29+-%3E+m+a+-%3E+m+a&scope=set%3Astackage

There are also versions which have a `m (Maybe a)` which I think corresponds to your current one. I think the version with more combinators is nicer.

j-mueller replied:
Yeah I added a `loopM` and used that to define the trigger that watches an address: https://github.com/input-output-hk/plutus/pull/1013/files#diff-0c70358a7c647e9fb38c726c94108085R55

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Event.hs:
```diff
@@ -0,0 +1,102 @@
+import           Ledger.Slot                          (Slot)
+import           Ledger.Tx                            (Address, Tx)
+
+data Step =
```
I don't like this name. Not sure what else to call it, though.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract.hs:
```diff
@@ -1,163 +1,38 @@
-    --       situation?
+-- | Watch an 'Address', returning the next transaction that changes it
+watchAddress :: Address -> PlutusContract Tx
+watchAddress a = await (Event.addr a) (fmap snd <$> ledgerUpdate) -- TODO: Check that tx actually involves the address
```
This will return the next transaction that changes *any* address we're watching, right?

j-mueller replied:
Fixed

michaelpj commented:
I'm suspicious of `applyInput`. It throws away inputs when it reaches e.g. `Pure`. So for example I think these two are not the same
```
(applyInput 2 (Pure 1)) >>= (\x -> Waiting (\i -> Pure i))
applyInput ((Pure 1) >>= (\x -> Waiting (\i -> Pure i))
```
Not sure those are exactly right, but you get the picture. We can bind into `Pure` computations later and provide things that can consume input, but those might have been thrown away by then.

I wonder if this is why `machines` has a separate `Plan` type. That prevents you from doing this "odd" interleaving of construction and evaluation.

michaelpj commented:
Might see if I can do this in terms of pipes.

michaelpj commented:
Part of an implementation using pipes: https://github.com/michaelpj/plutus/commit/74a5b85f2a8d63e38a71d4f509a876dbe06a105e

Not sure how viable this is, but I think it's somewhat promising.

vmchale commented:
For my edification, what is the advantage of pipes here?

michaelpj commented:
> For my edification, what is the advantage of pipes here?

We've got a similar structure here, better to use something that already exists. Plus I'm worried about getting the semantics right and I trust that most such issues have been ironed out of pipes. Plus conceptually we've got something that takes a stream of events and produces a stream of events, so pipes is actually quite a good fit.

I have even more of a worry, which is that these are supposed to be stateless, but I don't see how we can stop and resume something where the data structure contains a continuation without just replaying all the previous events?

j-mueller commented:
> without just replaying all the previous events?

I think we just have to replay all the previous events.

j-mueller commented:
> I'm suspicious of applyInput. It throws away inputs when it reaches e.g. Pure

Ah yes that's true! Maybe it should indicate whether that has happened.

michaelpj commented:
> I think we just have to replay all the previous events.

That's a little worrying. So startup time is going to be linear in the total history of the contract? Should be fast in many cases, but if you e.g. need to do some crypto to determine what to put in your tx, and you have a history of thousands...

j-mueller commented:
Thanks for the pipes version @michaelpj . I'll check it out now

j-mueller commented:
> That's a little worrying. So startup time is going to be linear in the total history of the contract? Should be fast in many cases, but if you e.g. need to do some crypto to determine what to put in your tx, and you have a history of thousands...

We can always build some kind of snapshot mechanism on top of that, as an additional endpoint. Also, re. the crypto, that shouldn't be a problem because of laziness - we only produce as output the last `Step` (and hence only the transactions that were produced in reaction to the last event)

michaelpj commented:
> We can always build some kind of snapshot mechanism on top of that, as an additional endpoint. Also, re. the crypto, that shouldn't be a problem because of laziness - we only produce as output the last Step (and hence only the transactions that were produced in reaction to the last event)

I'm unsure if that's true. Okay sure, if we produce a list in theory we could not compute the actual elements, but it would be very easy for something to force them, so I'd be wary of relying on it.

michaelpj commented:
Another thing: with the whole "await an event of the right kind", I think all our solutions have a problem if we have different things waiting for different events. At the moment if something looks at an event and it's not the one it wants it just discards it. I think we need some kind of parallel composition where inputs go to both sides. Not sure if pipes provides this.

michaelpj commented:
Possibly the semigroup instance does what we want, currently trying to understand it.

michaelpj commented:
I think it is not: you run the LHS until it halts and only then do you start running the RHS. So if the LHS consumes an input the RHS won't see it.

j-mueller commented:
I just ran `cabal new-run contract-exe-guessing-game` and `curl localhost:8080/initialise`, with the Writer change (https://github.com/michaelpj/plutus/commit/6a6539714a50513f17c93de554b57192ad31448a) and it gives a different result now.

6a6539714a50513f17c93de554b57192ad31448a (with writer):

```
{
  "stepEndpoints": [],
  "stepNextSlot": null,
  "stepAddresses": [],
  "stepTransactions": []
}
```

e4a30cd7fb96357548ab6580bd9ced416e3ea9ac (without writer):

```
{
  "stepEndpoints": [
    "lock"
  ],
  "stepNextSlot": null,
  "stepAddresses": [
    {
      "getAddress": "9f182f18e918f2185618270f18a41895187518f7182f18bc186b188a18541841041858183818d3185718291862188518fc1871188a1898187a18d6187e182fff"
    }
  ],
  "stepTransactions": []
}
```

The latter is how it is supposed to work, combining the expected inputs (`Step`s) of the [`guess`](https://github.com/input-output-hk/plutus/pull/1013/files#diff-015e10b145c7fce4f91a2c363578adb9R57) and [`lock`](https://github.com/input-output-hk/plutus/pull/1013/files#diff-015e10b145c7fce4f91a2c363578adb9R69) endpoints, ie. `main = run 8080 (contractApp (guess <> lock))`

j-mueller commented:
> I think we need some kind of parallel composition where inputs go to both sides

Yeah that was what I intended with the applicative instance, it does apply the input to both sides

michaelpj commented:
> I just ran cabal new-run contract-exe-guessing-game and curl localhost:8080/initialise, with the Writer change (michaelpj@6a65397) and it gives a different result now.

Hmmm, I just followed the types. I probably made a mistake, but I think the overall structure is more-right. I understand `applyInputs` as `foldM` whereas I have no idea what the current one does...

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,120 @@
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
```
Do we need `Done`?

j-mueller replied:
Probably not. But we might need an error / `Throw` case at some point

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,120 @@
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
```
Can we make it `Contract i o a`, I keep getting the parameters confused.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,120 @@
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
```
c.f. http://hackage.haskell.org/package/haxl-2.0.1.1/docs/src/Haxl.Core.Monad.html#line-569

Haxl is also doing the parallel applicative thing, and it has a similar structure.

j-mueller replied:
Yes I looked at haxl, but it's tied closely to the IO monad and geared towards concurrent execution and caching, which we don't need. The `concur` package also goes in a similar direction.

michaelpj replied:
Yeah, I just meant we can cross-check our implementation of stuff.

michaelpj commented:
Some tentative conclusions from a bunch of digging:
- Most of the Haskell streaming libraries are pull-based, which means they deal okay with coproducts but not products (can't request both sides give you an `a` at the same time). Products are fine in the push based model (can push an `a` to both sides at the same time).
- There doesn't seem to be a good push-based library. Maybe you can bend pipes into doing this, but I'm not sure.

So I'm coming around to writing our own, although ideally as close to existing stuff as possible.

michaelpj commented:
Maybe we want http://hackage.haskell.org/package/pipes-core-0.1.0/docs/Control-Pipe-Monoidal.html

j-mueller commented:
Going to add some tests now.

I came across some more issues that I hadn't considered before.

1. Sometimes we need to refer to a transaction that we submitted earlier. In the new model where we only produce `UnbalancedTx` transactions this is impossible, because the transaction ID (hash) is only known after coin selection has happened. I managed to work around that in the crowdfunding case, but we might have to add a callback for the app platform to inform the contract code of the tx ID of an `UnbalancedTx` after it has been finalised.
2. In the applicative instance we currently run both sides on every new event that has been received. This is fine for ledger events, but I think we need a different behaviour for endpoints - if the same endpoint happens to be active twice, for whatever reason, and the user activates it, then I think the intuitive behaviour is to only run one of the endpoint implementations in response. To resolve this we could either add a `WaitingAp` constructor that looks like `Waiting` but has the single-application behaviour for `<*>` (I don't think that would be very nice though). Or we could make sure that every endpoint is unique. For example by keeping track of the nesting of the endpoints when constructing the combined contract.
3. @mchakravarty had some concerns about the use of strings to identify endpoints. I don't think this is a problem (the types/schema of the endpoints are still well-defined) but still, maybe there is a better solution. A reason in favour of the current design is that it supports dynamic endpoints (the set of endpoints changes during the lifetime of a contract). In the playground all endpoints are static, which leads to situations where certain endpoints have to be run before or after other endpoints, otherwise the contract fails. IIRC Marlowe endpoints can also change between active and inactive. Dynamic endpoints would also allow us to report information about the contract back to the user. For example, how many contributions have been collected, or what the current value of a swap is. So, we need a dynamic endpoint mechnism that doesn't use `String`s to identify the endpoints. Maybe we could have a system that keeps track of all _possible_ endpoints at the type level, and then returns the names of the active endpoints after each step.

j-mueller commented:
After talking with @krisajenkins (and @mchakravarty earlier) I now think it's best to make all the contract endpoints known statically at compile time, and only change their status (active/inactive) during runtime. However, in the interest of getting something workable for the app platform integration I would prefer to merge this PR once the tests are done, and then make the endpoints static in another PR.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+select l r = 
+    case l of
+        Pure a -> Pure a
+        Emit e c -> Emit e (select c r)
```
Hah, I wonder if this was your bug, I was just about to comment that ignoring the emits seemed suspicious.

j-mueller replied:
yeah it might have had something to do with it ;)

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+
+-- | Run both contracts and return the first one that finishes
+select :: Contract i o a -> Contract i o a -> Contract i o a
+select l r = 
```
This function has a very similar shape to `<*>`, which makes me a bit suspicious, I wonder if it can be written in terms of it.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+result :: Monoid o => Contract i o a -> Maybe a
+result = (\case { Pure b -> Just b; _ -> Nothing }) . snd . drain
+
+await :: o -> (i -> Maybe a) -> Contract i o a
```
I remain suspicious about using this for triggers. As it is, all triggers must be set up in parallel, otherwise "downstream" triggers won't see any events, because they're discarded by earlier triggers looking at things. That seems like a very easy way to mess things up.

j-mueller replied:
> all triggers must be set up in parallel, otherwise "downstream" triggers won't see any events

That was the intention, the triggers aren't active all the time. If you use `select`, `until`, `both` etc. to build your contracts than it should work as expected.

michaelpj replied:
Hmm, so I think the situation I'm describing is something like (made up function names):
```
do
   _ <- waitUntilFundsAre x
   _ <- waitUntilSlotIs y
   doThing
````
Now, the question is: what happens if slot y passes and *then* funds reach x? I think if this was written in applicative style it would work, so probably using `ApplicativeDo` you're okay too. And if there is a data dependency (i.e. the second condition you're waiting on depends on the output of the first one, then of course it can't trigger until the first one does.

So I think this is indeed okay, although I'd like to see some tests of the basic mechanism witnessing the various patterns of awaiting etc.

j-mueller replied:
A general limitation of the current design is that there are no signals (in the FRP sense). So `waitUntilFundsAre` returns as soon as the funds are `x`, but then can't tell if the funds drop below `x` again by the time `waitUntilSlotIs` returns.

Although for your particular example you can use [`watchAddressUntil`](https://github.com/input-output-hk/plutus/pull/1013/files#diff-0c70358a7c647e9fb38c726c94108085R57).

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+    go b' = maybe (Left b') (Right . flip f b') <$> con
+
+-- | Apply an input to a contract
+applyInput
```
I think it would be simpler if if this was:
```
applyInput
    :: Monoid o
    => i
    -> Contract i o a
    -> Contract i o a
applyInput ip = \case
    Waiting f -> f ip
    Pure a -> Pure a
    Emit t c -> Emit t (applyInput ip c)
```

At the moment it's essentially my `applyInput` followed by `drain`. No need to fuse the two!

michaelpj replied:
In fact, then `applyInput` is  the `offer` that I wrote as an auxiliary to `<*>`. One difference: I didn't make `offer` apply inputs under `Emit`, which is probably wrong.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+    Pure a -> Pure a -- ??
+    Emit t c -> Emit t (applyInput ip c)
+
+applyInputs
```
Similarly, this doesn't need to collect outputs. Because we can always just go under the `Emit`s, we don't need to mix up applying inputs and draining.

michaelpj replied:
Also, can `applyInput` and `applyInputs` be consistent on whether the inputs go before or after the contract in the arguments?

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -0,0 +1,142 @@
+both :: Contract i o a -> Contract i o b -> Contract i o (a, b)
+both = liftA2 (,)
+
+firstOf :: Contract i o a -> Contract i o b -> Contract i o (Either a b)
```
`selectEither`?

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs:
```diff
@@ -36,15 +36,15 @@ instance Alt (Contract i o) where
+    Pure f    <*> ca       = f <$> ca
+    Waiting f <*> ca       = Waiting $ \i -> f i <*> offer i ca
+
+offer :: i -> Contract i o a -> Contract i o a
```
fwiw I think applyInput is a fine name, I just picked this not to clash :laughing: up to you

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Servant.hs:
```diff
@@ -25,7 +25,7 @@ contractServer c = initialise :<|> run where
     run (e:es) =
-      let c' = applyInputs c es in
+      let c' = applyInputs es c in
       pure $ fst $ drain $ applyInput e c'
```
Isn't this whole thing just `run es = pure $ fst $ drain $ applyInputs es c`?

j-mueller commented:
@michaelpj I think it's ready for proper review now. The notes `[Hooks and Events]` and `[Handling state in contracts]` describe the general idea, and the rest should follow from that :)

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Record.hs:
```diff
@@ -0,0 +1,69 @@
+    deriving stock (Eq, Show, Generic, Functor)
+    deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+
+data ClosedRecord i =
```
I think what you may want to do here is something like:
```
-- Whichever language extensions you need to get type promotion
{-# LANGUAGE DataKinds #-}

type Closure = Open | Closed

data Record c i where
    RecordLeaf :: i -> Closure c i
    RecordBothClosed :: Record Closed i -> Record Closed i -> Record Closed i
    RecordOpenLeft :: Record Open i -> record Closed i -> Record Open i
    ...
```
Possibly you could do something clever with type families to avoid the four cases there, but I'm not clever enough for that.

This *should* mean that if you pattern match on a `Record Closed i` you'll only get the leaf and both closed cases.
    ... -- other binary closed/open cases

j-mueller replied:
After working through the GADT code that we started in our pairing session it turned out that we always need to know which of the two branches was closed and which one was open, so there is no place where having a `Record Open i` actually reduces the number of cases we need to look at. So I reverted back to the previous version.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/State.hs:
```diff
@@ -0,0 +1,324 @@
+initialise
+    :: ( MonadWriter Hooks m )
+    => StatefulContract a
+    -> m (Either (OpenRecord Event) (ClosedRecord Event, a))
```
I agree that this type doesn't seem quite right, but I'm not entirely sure what else should be there.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/State.hs:
```diff
@@ -0,0 +1,324 @@
+checkpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+checkpoint = CJSONCheckpoint
+
+prtty :: StatefulContract  a -> String
```
One 'e' won't kill you ;)

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/State.hs:
```diff
@@ -0,0 +1,324 @@
+    CJSONCheckpoint j -> "json(" ++ prtty j ++ ")"
+
+instance Functor StatefulContract where
+    fmap f = \case
```
Couldn't this just be `fmap f = CMap f`?

j-mueller replied:
Yeah for some reason I wanted to keep the number of `CMap`s small, but it doesn't make a difference.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/State.hs:
```diff
@@ -0,0 +1,324 @@
+    pure = CContract . pure
+    (<*>) = CAp
+
+-- TODO: Should we add an `Alt` constructor to `StatefulContract`?
```
Don't we need one? I'd have thought that this would impede us from efficiently saving contracts defined via `<|>`, because they'll be lowered to `CContract`s so we'll only be able to serialize them when *all* the branches have reached a checkpoint. Whereas if there was a constructor I think we could do that.

j-mueller replied:
Yeah good point.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/State.hs:
```diff
@@ -0,0 +1,324 @@
+lower :: StatefulContract a -> C.ContractPrompt Maybe a
+lower = lowerM id id
+
+runConM
```
Should be in `Contract` perhaps?

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,171 @@
+-- | An unsigned and potentially unbalanced transaction, as produced by
+--   a contract endpoint. See note [Unbalanced transactions]
+data UnbalancedTx = UnbalancedTx
+        { _Inputs             :: [(L.TxIn, V.Value)]
```
Capitalizing these is weird.

michaelpj replied:
I take it that the reason this has values on it is so that you can calculate how much is needed to make it balance, but it's a bit odd. Do we actually need that ability? Can't we just hand it over to the wallet, which is presumably going to look stuff up in the index, etc?

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,171 @@
+
+-- | Combine two unbalanced transactions by appending their respective inputs,
+--   outputs, and signatures, adding their forged values, and applying the
+--   given function to their two validity ranges.
```
Seems to me that taking the intersection of the validity ranges is the only sensible thing to do. If we do that then this is sensibly a transaction that can stand in for both the provided transactions.

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,171 @@
+returns a 'Maybe Interval'.))
+
+TODO:  I think a good way forward would be to make
+'Interval' a lattice, and then make 'UnbalancedTx' a lattice too.
```
I strongly agree with making `Interval` a lattice. It's easy to add an empty interval (even using picking any interval with `l >h` would do).

I'm not sure that `UnbalancedTx` should be a lattice. The operator where you use the monoid on the other fields and `intersection` on the interval seems useful, and seems most naturally to be `join` (since it "does both"), I'm not sure what `meet` would be.

j-mueller replied:
> even using picking any interval with `l >h` would do

But how should we define `empty :: Interval a` ?

michaelpj replied:
Right, you probably need something like 
```
class Lattice a => BoundedLattice a where
    top :: a
    bottom :: a

instance BoundedLattice a => BoundedLattice (Interval a) where
    top = Interval bottom top
    bottom = Interval top bottom
````

Possibly we should make it possible to specify the closure of the endpoints in an interval since I think you probably want top to be `[bottom, top]` and bottom to be `(top, bottom)` so they work well in the degenerate case with one element.

j-mueller replied:
https://github.com/input-output-hk/plutus/issues/1309

j-mueller replied:
https://github.com/input-output-hk/plutus/issues/1309

michaelpj commented on plutus-contract-exe/src/Language/Plutus/Contract/Wallet.hs:
```diff
@@ -0,0 +1,90 @@
+    -- ^ The unbalanced transaction
+    -> m Tx
+balanceTx utxo pk tx =
+    let (neg, pos) = Value.split (T.computeBalance tx)
```
Yeah, so I think this should just require the `Index`, rather than expecting the unbalanced transaction to have the values in it.

j-mueller commented on plutus-contract-exe/src/Language/Plutus/Contract/Request.hs:
```diff
@@ -0,0 +1,56 @@
+    endpoints to mean any kind of capability that the contract may need (submit 
+    transactions, sign things, make HTTP calls, etc), not just user-facing 
+    endpoints. Then we can get rid of all other constructors in `Hook` and 
+    `Event`.
```
Yes I think that's correct. And I plan on using something that already exists when I get around to implementing #1115 .
What makes this effect system special is that all effects are variations of the `Prompt` effect and they only differ in what can be prompted for.

michaelpj replied:
We might want to hide that, though. From the point of view of the contract author they shouldn't really be able to tell whether "sign this" works by sending a message to the SCB and waiting for a response or signing it locally. So having a bunch of effects and just interpreting them all as prompts seems reasonable.

j-mueller replied:
I solved this by adding `extensible-effects` and encoding all interactions with the outside world as effects, see for example `Language.Plutus.Contract.Effects.AwaitSlot`. They get boiled down to a combination of the `State` and `Exc` effects under the hood.

j-mueller commented on plutus-contract-exe/src/Language/Plutus/Contract/Class.hs:
```diff
@@ -0,0 +1,41 @@
+foldMaybe f b con = loopM go b where
+    go b' = maybe (Right b') (Left . flip f b') <$> con
+
+await :: (Alternative m, Monad m, MonadContract i o m) => o -> (i -> Maybe a) -> m a
```
It needs to tell the outside world what it is waiting for (and the `prompt` library has the [same pattern](https://hackage.haskell.org/package/prompt-0.1.1.2/docs/Control-Monad-Prompt.html#v:prompt))

michaelpj replied:
`prompt` is a better name than `await`, I think.

j-mueller commented on plutus-contract-exe/src/Language/Plutus/Contract/Class.hs:
```diff
@@ -0,0 +1,41 @@
+await :: (Alternative m, Monad m, MonadContract i o m) => o -> (i -> Maybe a) -> m a
+await i f = do
+    o <- prompt i
+    maybe empty pure (f o)
```
Hm yeah `await` has been completely replaced by `prompt` so I'll delete it.

And you're right there used to be a `loopM` in there, but this behaviour (trying events on both branches of a parallel contract) is now achieved in [`runContract`](https://github.com/input-output-hk/plutus/pull/1013/files#diff-1f87d2978c1b16b45c43fa6373bbe57aR51)

j-mueller commented:
Alright I think it's ready. Notable changes since the last time are
* added `extensible-effects` to capture the effects of contracts (this necessitated the removal of the `PromptT` type), 
* Renamed `StatefulContract` to `Resumable` and made the whole `Language.Plutus.Contract.Resumable` module independent of the contract-specific types
* Added ` `Alt` and `Empty` constructors to `Resumable`
* Added `Language.Plutus.Contract.App.runWithTraces` which lets you run some contract-specific traces on the mockchain and print out the corresponding JSON values to be used as request bodies for calling the `run` HTTP endoint.

j-mueller commented:
Sorry that this got so large, I'll make smaller PRs in the future ;)

michaelpj commented:
I'll try and review this ASAP.

michaelpj commented:
Rogue `out.son` file.

michaelpj commented on plutus-contract/examples/lib/Examples/Crowdfunding.hs:
```diff
@@ -0,0 +1,76 @@
+
+scheduleCollection :: Plutus r => Campaign -> Contract r ()
+scheduleCollection cmp = do
+    () <- endpoint "schedule collection"
```
I had to stare at this for a while, then I remembered that "endpoint" means "wait for someone to call this endpoint and get the arguments". I think perhaps it could do with a better name.

michaelpj commented on plutus-contract/examples/lib/Examples/Crowdfunding.hs:
```diff
@@ -0,0 +1,76 @@
+    let trg = fst <$> both
+                (fundsAtAddressGt (CF.campaignAddress cmp) (campaignTarget cmp))
+                (awaitSlot (CF.campaignDeadline cmp))
+    void $ timeout (CF.campaignCollectionDeadline cmp) $ do
```
It's not totally obvious to me why we use two different mechanisms to await our refund and to wait for collection.

j-mueller replied:
In `scheduleCollection` we need to make sure that we don't try to collect any contributions before the `campaignDeadline` slot, but we also need to keep watching the address during that time, hence the `trg` action. If we used `watchAddressUntil` here, we would miss the case where contributions are discovered after `campaignDeadline` but before `campaignCollectionDeadline`.

michaelpj commented on plutus-contract/examples/lib/Examples/Game.hs:
```diff
@@ -0,0 +1,94 @@
+        vl         = Ada.toValue amt
+        dataScript = gameDataScript secret
+        output = L.TxOutOf gameAddress vl (L.PayToScript dataScript)
+        tx     = unbalancedTx [] [output]
```
It's very nice being able to do stuff like this.

michaelpj commented on plutus-contract/test/Spec/HUnit.hs:
```diff
@@ -0,0 +1,139 @@
+    -> TestTree
+checkPredicate nm con predicate action =
+    HUnit.testCase nm $
+        case runTrace con action of -- action of
```
commented out code

michaelpj commented on plutus-contract/test/Spec/Contract.hs:
```diff
@@ -0,0 +1,99 @@
+import           Spec.HUnit
+
+tests :: TestTree
+tests = testGroup "contracts"
```
These are very nice, well done!

michaelpj commented on plutus-contract/docs/contract-api.adoc:
```diff
@@ -0,0 +1,195 @@
+[#contract-api]
```
I guess this will eventually graduate to a tutorial chapter?

michaelpj commented on plutus-contract/docs/contract-api.adoc:
```diff
@@ -0,0 +1,195 @@
+[#contract-api]
+= Writing Plutus Contracts
+
+The `PlutusContract` type represents the client-side parts of Plutus contracts. `PlutusContract` values describe the interactions of contracts with the blockchain and with the user. They can be composed in different ways, allowing you to assemble contracts from the building blocks provided by the Plutus library and by other contract authors.
```
Seems to just be `Contract` now? This probably needs the LHS treatment to keep this stuff current.

j-mueller replied:
It will replace the plutus tutorial eventually: https://github.com/input-output-hk/plutus/issues/1321

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Wallet.hs:
```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+-- | Balance  `UnbalancedTx` values using the
+--   wallet API
```
Yes (this module is also relevant for the mock wallet backend @raduom )

michaelpj commented:
FYI you can add suggestions to a batch and commit them all together, you don't have to do one commit for each ;)

j-mueller commented:
hm I was looking for that feature but couldn't find it :( anway, it's done now.

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Effects.hs:
```diff
@@ -0,0 +1,100 @@
+
+* `Event` values are produced by the app platform. Each constructor of `Event`
+  corresponds to a constructor of `Hook a`. For example, `LedgerUpdate ::
+  Address -> Tx -> Event` is the counterpart of `AddrHook`.
```
Yes ideally we should have an open union of request-response pairs, one for each of the effects of the contract. I'll leave that to https://github.com/input-output-hk/plutus/issues/1317 though (all our contracts have the same set of effect so, even though it would be nice to have now, we really only need it when we have other effects or different backends)

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Effects.hs:
```diff
@@ -0,0 +1,100 @@
+
+TODO: Use the type parameter of `Hooks` to encode the type of endpoints that a
+   contract may expose. Once we can do that, we can generalise the notion of
+    endpoints to mean any kind of capability that the contract may need (submit
```
Yeah the second part of that comment is outdated, now that we have `extensible-effects` for those things. I'll update it.

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,161 @@
+-- | The ledger transaction of the 'UnbalancedTx'. Note that the result
+--   does not have any signatures, and is potentially unbalanced (ie. invalid).
+--   To produce a balanced 'Tx', use 'Language.Plutus.Contract.Wallet.balance'.
+toLedgerTx :: UnbalancedTx -> L.Tx
```
I don't think this would give us anything. We likely have to add inputs and outputs to make it balanced before we can sign it, so adding the signatures here would only make sense if it was balanced & paid up already. Perhaps we can make this nicer with better types for transactions: https://github.com/input-output-hk/plutus/issues/1320


 (Also the lookup function would have to be `(PubKey -> L.Tx -> Maybe Signature)`)

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,161 @@
+collectFromScript = collectFromScriptFilter (\_ -> const True)
+
+-- | See 'Wallet.API.collectFromScriptFilter'
+collectFromScriptFilter
```
https://github.com/input-output-hk/plutus/issues/1320

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Transaction.hs:
```diff
@@ -0,0 +1,161 @@
+interval. There are two valid options: Hull and intersection. ('always' is the
+unit for the intersection but then there is the issue that we don't have a
+canonical representation of the empty interval (that's why 'intersection'
+returns a 'Maybe Interval'.))
```
https://github.com/input-output-hk/plutus/issues/1309

j-mueller replied:
https://github.com/input-output-hk/plutus/issues/1309

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Wallet.hs:
```diff
@@ -0,0 +1,119 @@
+    -> m Tx
+balanceTx utxo pk tx = do
+    (neg, pos) <- Value.split <$> computeBalance tx
+    let tx' = addOutputs pk pos (T.toLedgerTx tx)
```
We could merge all pubkey outputs that go to the same key into a single one. But I'm not sure if that is always desirable - if multiple currencies are involved then we might want to keep them on separate outputs to reduce the number of change outputs we need in the future.

michaelpj replied:
Yes, that should probably be configurable on the user's end. Having an extra output for the (possibly empty) excess on the original balance seems less good, though.

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Emulator.hs:
```diff
@@ -0,0 +1,222 @@
+
+data ContractTraceState a =
+    ContractTraceState
+        { _ctsEvents   :: Map Wallet (Seq Event)
```
We can always recover `[Record]` values from the `[Event]` but not the other way round

j-mueller commented on plutus-contract/src/Language/Plutus/Contract.hs:
```diff
@@ -0,0 +1,127 @@
+
+-- | Run a contract until the given slot has been reached.
+until :: (Member AwaitSlot r) => Contract r a -> Slot -> Contract r (Maybe a)
+until c sl = fmap (either (const Nothing) Just) (selectEither (awaitSlot sl) c)
```
:(

michaelpj replied:
Up to you, just my style :)

effectfully replied:
I'd write `reoption <$> selectEither (awaitSlot sl) c` ([`reoption`](https://github.com/input-output-hk/plutus/blob/8ea96a114a7bc800abfa2e2005baa3dced4115ae/language-plutus-core/prelude/PlutusPrelude.hs#L232) is defined in `PlutusPrelude`).

j-mueller commented on plutus-contract/src/Language/Plutus/Contract/Resumable.hs:
```diff
@@ -0,0 +1,348 @@
+--   asking for a different input value. 'Step i o' is like an inverted request-
+--   response pair: We offer the response (in form of 'Maybe i'), and only if it
+--   is the wrong response do we record the request in form of 'o'. 
+newtype Step i o a = Step { runStep :: Maybe i -> Either o a }
```
Yeah it's precisely for `initialise` that we need the `Maybe Input`, just in case the step doesn't actually need any input (if it's `pure a`)

michaelpj replied:
I guess what I'm saying is perhaps we could write the functions more generically, and then the type of "functions that can be initialized" is `Step (Maybe i) o a` or something.

j-mueller commented:
@michaelpj I think I've addressed everything

michaelpj commented:
This has been hanging around for long enough, let's get it going.
